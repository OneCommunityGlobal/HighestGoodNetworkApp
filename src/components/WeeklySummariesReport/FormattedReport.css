.star-rating {
    position: relative;
    display: inline-block;
  }
  
  .rating-number {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    text-align: center;
    font-size: 0.8em;
    line-height: 4.0;
    color: azure;
  }
  