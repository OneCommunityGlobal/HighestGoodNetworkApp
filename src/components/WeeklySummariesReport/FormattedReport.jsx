import React, { useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import 'moment-timezone';
import ReactHtmlParser from 'react-html-parser';
import { Link } from 'react-router-dom';
import google_doc_icon from './google_doc_icon.png';
import './WeeklySummariesReport.css';
import { toast } from 'react-toastify';
import ToggleSwitch from '../UserProfile/UserProfileEdit/ToggleSwitch';
import axios from 'axios';
import { ENDPOINTS } from '../../utils/URL';

import { assignStarDotColors, showStar } from 'utils/leaderboardPermissions';

const textColors = {
  Default: '#000000',
  'Not Required': '#708090',
  Team: '#FF00FF',
  'Team Fabulous': '#FF00FF',
  'Team Marigold': '#FF7F00',
  'Team Luminous': '#C4AF18',
  'Team Lush': '#00FF00',
  'Team Sky': '#0000FF',
  'Team Azure': '#4B0082',
  'Team Amethyst': '#9400D3',
};

const FormattedReport = ({ summaries, weekIndex, bioCanEdit }) => {
  const emails = [];
  //const bioCanEdit = role === 'Owner' || role === 'Administrator';

  summaries.forEach(summary => {
    if (summary.email !== undefined && summary.email !== null) {
      emails.push(summary.email);
    }
  });

  //Necessary because our version of node is outdated
  //and doesn't have String.prototype.replaceAll
  let emailString = [...new Set(emails)].toString();
  while (emailString.includes(',')) emailString = emailString.replace(',', '\n');
  while (emailString.includes('\n')) emailString = emailString.replace('\n', ', ');

  const alphabetize = summaries => {
    const temp = [...summaries];
    return temp.sort((a, b) =>
      `${a.firstName} ${a.lastName}`.localeCompare(`${b.firstName} ${b.lastname}`),
    );
  };

  const getMediaUrlLink = summary => {
    if (summary.mediaUrl) {
      return (
        <a href={summary.mediaUrl} target="_blank" rel="noopener noreferrer">
          Open link to media files
        </a>
      );
    } else {
      return 'Not provided!';
    }
  };

  const getGoogleDocLink = summary => {
    if (!summary.adminLinks) {
      return undefined;
    }

    const googleDocLink = summary.adminLinks.find(link => link.Name === 'Google Doc');

    return googleDocLink;
  };

  const getWeeklySummaryMessage = summary => {
    if (!summary) {
      return (
        <p>
          <b>Weekly Summary:</b> Not provided!
        </p>
      );
    }

    const summaryText = summary?.weeklySummaries[weekIndex]?.summary;
    let summaryDate = moment()
      .tz('America/Los_Angeles')
      .endOf('week')
      .subtract(weekIndex, 'week')
      .format('YYYY-MMM-DD');
    let summaryDateText = `Weekly Summary (${summaryDate}):`;
    const summaryContent = (() => {
      if (summaryText) {
        const style = {
          color: textColors[summary?.weeklySummaryOption] || textColors['Default'],
        };

        summaryDate = moment(summary.weeklySummaries[weekIndex]?.uploadDate)
          .tz('America/Los_Angeles')
          .format('YYYY-MMM-DD');
        summaryDateText = `Summary Submitted On (${summaryDate}):`;

        return <div style={style}>{ReactHtmlParser(summaryText)}</div>;
      } else {
        if (
          summary?.weeklySummaryOption === 'Not Required' ||
          (!summary?.weeklySummaryOption && summary.weeklySummaryNotReq)
        ) {
          return <p style={{ color: textColors['Not Required'] }}>Not required for this user</p>;
        } else {
          return <span style={{ color: 'red' }}>Not provided!</span>;
        }
      }
    })();

    return (
      <>
        <p>
          <b>{summaryDateText}</b>
        </p>
        {summaryContent}
      </>
    );
  };

  const getTotalValidWeeklySummaries = summary => {
    const style = {
      color: textColors[summary?.weeklySummaryOption] || textColors['Default'],
    };
    if (summary.weeklySummariesCount === 8) {
      return (
        <p>
          <b style={style}>Total Valid Weekly Summaries:</b>{' '}
          {summary.weeklySummariesCount || 'No valid submissions yet!'}
        </p>
      );
    } else {
      return (
        <p>
          <b style={style}>Total Valid Weekly Summaries:</b>{' '}
          {summary.weeklySummariesCount || 'No valid submissions yet!'}
        </p>
      );
    }
  };

  const handleGoogleDocClick = googleDocLink => {
    const toastGoogleLinkDoesNotExist = 'toast-on-click';
    if (googleDocLink) {
      window.open(googleDocLink.Link);
    } else {
      toast.error(
        'Uh oh, no Google Doc is present for this user! Please contact an Admin to find out why.',
        {
          toastId: toastGoogleLinkDoesNotExist,
          pauseOnFocusLoss: false,
          autoClose: 3000,
        },
      );
    }
  };

  const handleChangeBioPosted = async (userId, bioStatus) => {
    try {
      const url = ENDPOINTS.USER_PROFILE(userId);
      const response = await axios.get(url);
      const userProfile = response.data;
      const res = await axios.put(url, {
        ...userProfile,
        bioPosted: bioStatus,
      });
      if (res.status === 200) {
        toast.success('You have changed the bio announcement status of this user.');
      }
    } catch (err) {
      alert('An error occurred while attempting to save the bioPosted change to the profile.');
    }
  };

  const BioSwitch = (userId, bioPosted, summary) => {
    const [bioStatus, setBioStatus] = useState(bioPosted);
    const style = {
      color: textColors[summary?.weeklySummaryOption] || textColors['Default'],
    };
    return (
      <div>
        <div className="bio-toggle">
          <b style={style}>Bio announcement:</b>
        </div>
        <div className="bio-toggle">
          <ToggleSwitch
            switchType="bio"
            state={bioStatus}
            handleUserProfile={bio => {
              setBioStatus(bio);
              handleChangeBioPosted(userId, bio);
            }}
          />
        </div>
      </div>
    );
  };

  const BioLabel = (userId, bioPosted, summary) => {
    const style = {
      color: textColors[summary?.weeklySummaryOption] || textColors['Default'],
    };
    return (
      <div>
        <b style={style}>Bio announcement:</b>
        {bioPosted === 'default'
          ? ' Not requested/posted'
          : bioPosted === 'posted'
          ? ' Posted'
          : ' Requested'}
      </div>
    );
  };

  const bioFunction = bioCanEdit ? BioSwitch : BioLabel;

  return (
    <>
      {alphabetize(summaries).map((summary, index) => {
        const hoursLogged = (summary.totalSeconds[weekIndex] || 0) / 3600;
        const googleDocLink = getGoogleDocLink(summary);
        return (
          <div
            style={{ padding: '20px 0', marginTop: '5px', borderBottom: '1px solid #DEE2E6' }}
            key={'summary-' + index}
          >
            <div>
              <b>Name: </b>
              <Link to={`/userProfile/${summary._id}`} title="View Profile">
                {summary.firstName} {summary.lastName}
              </Link>

              <span onClick={() => handleGoogleDocClick(googleDocLink)}>
                <img className="google-doc-icon" src={google_doc_icon} alt="google_doc" />
              </span>
              <span>
                <b>&nbsp;&nbsp;{summary.role !== 'Volunteer' && `(${summary.role})`}</b>
              </span>
              {showStar(hoursLogged, summary.weeklycommittedHours) && (
                <i
                  className="fa fa-star"
                  title={`Weekly Committed: ${summary.weeklycommittedHours} hours`}
                  style={{
                    color: assignStarDotColors(hoursLogged, summary.weeklycommittedHours),
                    fontSize: '55px',
                    marginLeft: '10px',
                    verticalAlign: 'middle',
                    position: 'relative',
                  }}
                >
                  <span
                    style={{
                      position: 'absolute',
                      top: '50%',
                      left: '50%',
                      transform: 'translate(-50%, -50%)',
                      color: 'white',
                      fontWeight: 'bold',
                      fontSize: '10px',
                    }}
                  >
                    +{Math.round((hoursLogged / summary.weeklycommittedHours - 1) * 100)}%
                  </span>
                </i>
              )}
            </div>
            <div>
              {' '}
              <b>Media URL:</b> {getMediaUrlLink(summary)}
            </div>
            {bioFunction(summary._id, summary.bioPosted, summary)}
            {getTotalValidWeeklySummaries(summary)}
            {hoursLogged >= summary.weeklycommittedHours && (
              <p>
                <b
                  style={{
                    color: textColors[summary?.weeklySummaryOption] || textColors['Default'],
                  }}
                >
                  Hours logged:{' '}
                </b>
                {hoursLogged.toFixed(2)} / {summary.weeklycommittedHours}
              </p>
            )}
            {hoursLogged < summary.weeklycommittedHours && (
              <p>
                <b
                  style={{
                    color: textColors[summary?.weeklySummaryOption] || textColors['Default'],
                  }}
                >
                  Hours logged:
                </b>{' '}
                {hoursLogged.toFixed(2)} / {summary.weeklycommittedHours}
              </p>
            )}
            {getWeeklySummaryMessage(summary)}
          </div>
        );
      })}
      <h4>Emails</h4>
      <p>{emailString}</p>
    </>
  );
};

FormattedReport.propTypes = {
  summaries: PropTypes.arrayOf(PropTypes.object).isRequired,
  weekIndex: PropTypes.number.isRequired,
};

export default FormattedReport;
