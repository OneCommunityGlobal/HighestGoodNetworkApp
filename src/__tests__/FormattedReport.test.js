import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import mockSummaries from 'weeklySummariesReportData'; // Located in the tested component's __mocks__ folder
import FormattedReport from '../components/WeeklySummariesReport/FormattedReport';

describe('FormattedReport Component', () => {
  it('Snapshot with mocked data', () => {
    // const { asFragment } = render(<FormattedReport summaries={mockSummaries} weekIndex="1" />);
    // expect(asFragment()).toMatchSnapshot();
  });
});
