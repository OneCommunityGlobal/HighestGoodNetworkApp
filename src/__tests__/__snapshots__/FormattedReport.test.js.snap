// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedReport Component Snapshot with mocked data 1`] = `
<DocumentFragment>
  <div
    style="padding: 20px 0px; margin-top: 5px; border-bottom: 1px solid #DEE2E6;"
  >
    <p>
      <b>
        Name:
      </b>
       A B
    </p>
    <p>
       
      <b>
        Media URL:
      </b>
       
      <a
        href="a"
        rel="noopener noreferrer"
        target="_blank"
      >
        Open link to media files
      </a>
    </p>
    <p>
      <b>
        Total Valid Weekly Summaries:
      </b>
       3
    </p>
    <p>
      <b>
        Committed weekly hours:
      </b>
       
    </p>
    <p>
      <b>
        Weekly Summary
      </b>
       (2020-Jun-27):
    </p>
    Week 2 summary
  </div>
  <div
    style="padding: 20px 0px; margin-top: 5px; border-bottom: 1px solid #DEE2E6;"
  >
    <p>
      <b>
        Name:
      </b>
       C D
    </p>
    <p>
       
      <b>
        Media URL:
      </b>
       Not provided!
    </p>
    <p>
      <b>
        Total Valid Weekly Summaries:
      </b>
       1
    </p>
    <p>
      <b>
        Committed weekly hours:
      </b>
       
    </p>
    <p>
      <b>
        Weekly Summary
      </b>
       (2020-Jun-27):
    </p>
    <span
      style="color: red;"
    >
      Not provided!
    </span>
  </div>
  <div
    style="padding: 20px 0px; margin-top: 5px; border-bottom: 1px solid #DEE2E6;"
  >
    <p>
      <b>
        Name:
      </b>
       E F
    </p>
    <p>
       
      <b>
        Media URL:
      </b>
       Not provided!
    </p>
    <p>
      <b>
        Total Valid Weekly Summaries:
      </b>
       No valid submissions yet!
    </p>
    <p>
      <b>
        Committed weekly hours:
      </b>
       
    </p>
    <p>
      <b>
        Weekly Summary
      </b>
       (2021-Aug-23):
    </p>
    <span
      style="color: red;"
    >
      Not provided!
    </span>
  </div>
  <h4>
    Emails
  </h4>
  <p />
</DocumentFragment>
`;
